File: vps-bot/python/api_client.py
from __future__ import annotations

import requests
from config import BACKEND_URL, BOT_API_SECRET

_HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {BOT_API_SECRET}",
}
_TIMEOUT = 15


def _get(path: str, params: dict | None = None) -> dict:
    resp = requests.get(f"{BACKEND_URL}{path}", params=params, headers=_HEADERS, timeout=_TIMEOUT)
    resp.raise_for_status()
    return resp.json()


def _post(path: str, body: dict | None = None) -> dict:
    resp = requests.post(f"{BACKEND_URL}{path}", json=body or {}, headers=_HEADERS, timeout=_TIMEOUT)
    resp.raise_for_status()
    return resp.json()


def get_pending_jobs(game: str) -> list[dict]:
    data = _get("/api/bot/jobs/pending", params={"game": game})
    return data.get("jobs", [])


def start_job(job_id: str) -> dict:
    return _post(f"/api/bot/jobs/{job_id}/start")


def trade_detected(job_id: str, detected_items: list[dict], proof_text: str | None = None) -> dict:
    return _post(f"/api/bot/jobs/{job_id}/trade-detected", {
        "detectedItems": detected_items,
        "proofText": proof_text,
    })


def complete_job(job_id: str, detected_items: list[dict], proof_text: str | None = None) -> dict:
    return _post(f"/api/bot/jobs/{job_id}/complete", {
        "detectedItems": detected_items,
        "proofText": proof_text,
        "requireDetection": True,
    })


def fail_job(job_id: str, error: str, retryable: bool = False) -> dict:
    return _post(f"/api/bot/jobs/{job_id}/fail", {
        "error": error,
        "retryable": retryable,
    })

File: vps-bot/src/api-client.ts
import { config } from "./config";

const BASE_HEADERS = {
  "Content-Type": "application/json",
  Authorization: `Bearer ${config.botApiSecret}`,
};

async function apiFetch<T = unknown>(
  path: string,
  init?: RequestInit
): Promise<T> {
  const url = `${config.backendUrl}${path}`;
  const res = await fetch(url, {
    ...init,
    headers: { ...BASE_HEADERS, ...(init?.headers as Record<string, string>) },
  });
  if (!res.ok) {
    const text = await res.text().catch(() => res.statusText);
    throw new Error(`[api] ${init?.method ?? "GET"} ${path} → ${res.status}: ${text}`);
  }
  return res.json() as Promise<T>;
}

export interface WaitingEntry {
  jobId: string;
  withdrawalId: string;
  customerRobloxUsername: string;
  botRobloxUsername: string | null;
  game: string;
  withdrawalStatus: "WAITING_FRIEND_REQUEST" | "WAITING_JOIN";
}

export const api = {
  heartbeat(status: "IDLE" | "ACTIVE" | "BUSY" | "ERROR") {
    return apiFetch("/api/bot/heartbeat", {
      method: "POST",
      body: JSON.stringify({
        botRobloxUsername: config.botUsername,
        game: config.game,
        status,
      }),
    });
  },

  getWaitingUsernames(): Promise<{ waiting: WaitingEntry[] }> {
    return apiFetch("/api/bot/waiting-usernames");
  },

  friendAccepted(customerRobloxUsername: string) {
    return apiFetch("/api/bot/friend-accepted", {
      method: "POST",
      body: JSON.stringify({
        customerRobloxUsername,
        botRobloxUsername: config.botUsername,
      }),
    });
  },

  customerJoined(jobId: string) {
    return apiFetch(`/api/bot/jobs/${jobId}/customer-joined`, {
      method: "POST",
    });
  },

  expireStaleOrders(timeoutHours: number) {
    return apiFetch<{ expired: number; total: number; withdrawalCodes: string[] }>(
      "/api/bot/expire-stale-orders",
      {
        method: "POST",
        body: JSON.stringify({ timeoutHours }),
      }
    );
  },
};
